File: plugins/lyy-dev-plugin/skills/java-code-review/scripts/git_diff.py
"""通用 Git Diff 包装脚本

执行 git diff {target}...{source} -- <pathspecs>，输出原始 diff 结果。
供 Java Code Review Skill 的 Agent 2/3/4 统一调用。

用法:
    python git_diff.py <repo> --source <branch> --target <branch> [-- pathspec ...]

示例:
    # Agent 2: Java 文件（排除测试）
    python git_diff.py /repo --source feat/x --target origin/master -- "*.java" ":(exclude)*/src/test/*"

    # Agent 3: 配置文件（排除 java/xml/md）
    python git_diff.py /repo --source feat/x --target origin/master -- ":(exclude)*.java" ":(exclude)*.xml" ":(exclude)*.md"

    # Agent 4: XML 文件（排除 pom.xml）
    python git_diff.py /repo --source feat/x --target origin/master -- "*.xml" ":(exclude)*pom.xml"

    # 无 pathspec，获取全部变更
    python git_diff.py /repo --source feat/x --target origin/master
"""

import argparse
import os
import subprocess
import sys


def _decode(data: bytes) -> str:
    for enc in ("utf-8", "gbk"):
        try:
            return data.decode(enc)
        except (UnicodeDecodeError, LookupError):
            continue
    return data.decode(sys.getdefaultencoding(), errors="replace")


def _validate_repo(path: str) -> str:
    abs_path = os.path.abspath(path)
    if not os.path.isdir(abs_path):
        raise SystemExit(f"路径不存在或不是目录: {abs_path}")
    if not os.path.exists(os.path.join(abs_path, ".git")):
        raise SystemExit(f"不是有效的 Git 仓库: {abs_path}")
    return abs_path


def _validate_branch(repo: str, branch: str) -> None:
    proc = subprocess.run(
        ["git", "-C", repo, "rev-parse", "--verify", branch],
        stdout=subprocess.PIPE, stderr=subprocess.PIPE,
    )
    if proc.returncode != 0:
        raise SystemExit(f"分支不存在: {branch}\n  {_decode(proc.stderr).strip()}")


def git_diff(repo: str, source: str, target: str, pathspecs: list) -> str:
    """执行 git diff {target}...{source} -- [pathspecs] 并返回原始输出。"""
    cmd = ["git", "-C", repo, "diff", f"{target}...{source}"]
    if pathspecs:
        cmd.append("--")
        cmd.extend(pathspecs)

    proc = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    if proc.returncode != 0:
        stderr = _decode(proc.stderr).strip()
        raise SystemExit(f"git diff 执行失败:\n  {stderr}")
    return _decode(proc.stdout)


def main() -> None:
    # 手动分割 -- 前后的参数，-- 后的部分作为 pathspec
    argv = sys.argv[1:]
    if "--" in argv:
        idx = argv.index("--")
        script_args = argv[:idx]
        pathspecs = argv[idx + 1:]
    else:
        script_args = argv
        pathspecs = []

    parser = argparse.ArgumentParser(description="通用 Git Diff 包装脚本")
    parser.add_argument("repo", help="Git 仓库路径")
    parser.add_argument("--source", required=True, help="源分支")
    parser.add_argument("--target", default="origin/master", help="目标分支（默认 origin/master）")
    args = parser.parse_args(script_args)

    repo = _validate_repo(args.repo)
    _validate_branch(repo, args.source)
    _validate_branch(repo, args.target)

    output = git_diff(repo, args.source, args.target, pathspecs)
    if output:
        print(output, end="")
    else:
        print("无变更内容。", file=sys.stderr)
        sys.exit(0)


if __name__ == "__main__":
    main()

File: plugins/lyy-dev-plugin/skills/java-code-review/scripts/git_diff.mjs
/**
 * 通用 Git Diff 包装脚本
 *
 * 执行 git diff {target}...{source} -- <pathspecs>，输出原始 diff 结果。
 * 供 Java Code Review Skill 的 Agent 2/3/4 统一调用。
 *
 * 用法:
 *   node git_diff.mjs <repo> --source <branch> --target <branch> [-- pathspec ...]
 *
 * 示例:
 *   # Agent 2: Java 文件（排除测试）
 *   node git_diff.mjs /repo --source feat/x --target origin/master -- "*.java" ":(exclude)*\/src/test/*"
 *
 *   # Agent 3: 配置文件（排除 java/xml/md）
 *   node git_diff.mjs /repo --source feat/x --target origin/master -- ":(exclude)*.java" ":(exclude)*.xml" ":(exclude)*.md"
 *
 *   # Agent 4: XML 文件（排除 pom.xml）
 *   node git_diff.mjs /repo --source feat/x --target origin/master -- "*.xml" ":(exclude)*pom.xml"
 *
 *   # 无 pathspec，获取全部变更
 *   node git_diff.mjs /repo --source feat/x --target origin/master
 */

import { execFileSync } from "node:child_process";
import { resolve, isAbsolute } from "node:path";
import { statSync } from "node:fs";

// ---------------------------------------------------------------------------
// 参数解析
// ---------------------------------------------------------------------------

function parseArgs(argv) {
  // 分割 -- 前后的参数
  const ddIdx = argv.indexOf("--");
  const scriptArgs = ddIdx >= 0 ? argv.slice(0, ddIdx) : argv;
  const pathspecs = ddIdx >= 0 ? argv.slice(ddIdx + 1) : [];

  let repo = null;
  let source = null;
  let target = "origin/master";

  for (let i = 0; i < scriptArgs.length; i++) {
    if (scriptArgs[i] === "--source" && i + 1 < scriptArgs.length) {
      source = scriptArgs[++i];
    } else if (scriptArgs[i] === "--target" && i + 1 < scriptArgs.length) {
      target = scriptArgs[++i];
    } else if (!repo && !scriptArgs[i].startsWith("--")) {
      repo = scriptArgs[i];
    }
  }

  if (!repo) {
    console.error("错误: 缺少必需参数 <repo>");
    process.exit(1);
  }
  if (!source) {
    console.error("错误: 缺少必需参数 --source");
    process.exit(1);
  }

  return { repo, source, target, pathspecs };
}

// ---------------------------------------------------------------------------
// 验证函数
// ---------------------------------------------------------------------------

function validateRepo(repoPath) {
  const absPath = isAbsolute(repoPath) ? repoPath : resolve(repoPath);
  try {
    const stat = statSync(absPath);
    if (!stat.isDirectory()) {
      console.error(`路径不是目录: ${absPath}`);
      process.exit(1);
    }
  } catch {
    console.error(`路径不存在: ${absPath}`);
    process.exit(1);
  }

  try {
    execFileSync("git", ["-C", absPath, "rev-parse", "--git-dir"], {
      stdio: ["pipe", "pipe", "pipe"],
    });
  } catch {
    console.error(`不是有效的 Git 仓库: ${absPath}`);
    process.exit(1);
  }

  return absPath;
}

function validateBranch(repo, branch) {
  try {
    execFileSync("git", ["-C", repo, "rev-parse", "--verify", branch], {
      stdio: ["pipe", "pipe", "pipe"],
    });
  } catch (err) {
    const stderr = err.stderr ? err.stderr.toString().trim() : "";
    console.error(`分支不存在: ${branch}\n  ${stderr}`);
    process.exit(1);
  }
}

// ---------------------------------------------------------------------------
// Git diff 执行
// ---------------------------------------------------------------------------

function gitDiff(repo, source, target, pathspecs) {
  const cmd = ["-C", repo, "diff", `${target}...${source}`];
  if (pathspecs.length > 0) {
    cmd.push("--");
    cmd.push(...pathspecs);
  }

  try {
    const stdout = execFileSync("git", cmd, {
      maxBuffer: 50 * 1024 * 1024, // 50 MB
    });
    return stdout.toString("utf-8");
  } catch (err) {
    const stderr = err.stderr ? err.stderr.toString().trim() : err.message;
    console.error(`git diff 执行失败:\n  ${stderr}`);
    process.exit(1);
  }
}

// ---------------------------------------------------------------------------
// 主流程
// ---------------------------------------------------------------------------

const args = parseArgs(process.argv.slice(2));
const repo = validateRepo(args.repo);
validateBranch(repo, args.source);
validateBranch(repo, args.target);

const output = gitDiff(repo, args.source, args.target, args.pathspecs);
if (output) {
  process.stdout.write(output);
} else {
  console.error("无变更内容。");
}
